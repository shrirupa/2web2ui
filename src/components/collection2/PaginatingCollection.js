import React, { Component, Fragment } from 'react';
import qs from 'query-string';
import { withRouter } from 'react-router-dom';
import TableCollectionView from './TableCollectionView';
import Pagination from './Pagination';

export class PaginatingCollection extends Component {
  state = {}

  static defaultProps = {
    defaultPerPage: 25
  }

  updateRows() {
    const { currentPage, perPage } = this.state;
    this.setState({ rows: null });
    return Promise.resolve(this.props.fetchRows({ currentPage, perPage }))
      .then(({ rows, rowCount }) => this.setState({ rows, rowCount }));
  }

  pageParamsChanged() {
    this.maybeUpdateQueryString();
    this.updateRows();
  }

  static getDerivedStateFromProps(props, state) {
    if (state.rows) {
      return null;
    }

    // If we're given rows on mount, use those to start with
    return {
      rows: props.rows,
      rowCount: props.rowCount || (props.rows ? props.rows.length : 0),
      perPage: props.defaultPerPage,
      currentPage: Number(qs.parse(props.location.search).page) || 1
    };
  }

  componentDidMount() {
    if (!this.state.rows) {
      // Load rows if none provided by the parent
      this.updateRows();
    }
  }

  handlePageChange = (index) => {
    const currentPage = index + 1;
    // Guard against Matchbox's Pagination component calling onChange on mount.
    if (this.state.currentPage !== currentPage) {
      this.setState({ currentPage }, this.pageParamsChanged);
    }
  }

  handlePerPageChange = (perPage) => {
    this.setState({ perPage, currentPage: 1 }, this.pageParamsChanged);
  }

  maybeUpdateQueryString() {
    const { location, updateQueryString } = this.props;
    if (updateQueryString === false) {
      return;
    }
    const { currentPage, perPage } = this.state;
    const { search, pathname } = location;
    const parsed = qs.parse(search);
    if (parsed.page || updateQueryString) {
      const updated = Object.assign(parsed, { page: currentPage, perPage });
      this.props.history.push(`${pathname}?${qs.stringify(updated)}`);
    }
  }

  renderPagination() {
    const { perPageButtons } = this.props;
    const { rowCount, currentPage, perPage } = this.state;

    return <Pagination
      saveCsv={false}
      rowCount={rowCount}
      perPage={perPage}
      currentPage={currentPage}
      perPageButtons={perPageButtons}
      onPageChange={this.handlePageChange}
      onPerPageChange={this.handlePerPageChange}
    />;
  }

  render() {
    const { collectionComponent: Coll = TableCollectionView } = this.props;
    return <Fragment>
      <Coll {...this.props} rows={this.state.rows} />
      {this.renderPagination()}
    </Fragment>;
  }
}

export default withRouter(PaginatingCollection);
