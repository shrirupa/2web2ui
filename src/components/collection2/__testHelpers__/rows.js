export function mkRow(i = 0) {
  return { id: i, name: `row-name-${Math.floor(i / 5)}-${i}` };
}

export function mkRows(n) {
  const rows = [];
  for (let i = 0; i < n; i++) {
    rows.push(mkRow(i));
  }
  return rows;
}
