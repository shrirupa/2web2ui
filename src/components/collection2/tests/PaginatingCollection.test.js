import React from 'react';
import { shallow } from 'enzyme';
import { PaginatingCollection } from '../PaginatingCollection';
import { mkRows } from '../__testHelpers__/rows';

describe('Component: PaginatingCollection', () => {
  let props;
  let wrapper;
  let rows;

  beforeEach(() => {
    rows = mkRows(10);
    props = {
      fetchRows: jest.fn().mockReturnValue({ rows, rowCount: rows.length }),
      defaultPerPage: 5,
      rows,
      location: {
        search: '',
        pathname: '/'
      },
      history: {
        push: jest.fn()
      }
    };
    wrapper = shallow(<PaginatingCollection {...props} />);
  });

  it('should render initial rows if provided', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should fetch rows on mount if not provided', () => {
    delete props.rows;
    shallow(<PaginatingCollection {...props} />);
    expect(props.fetchRows).toHaveBeenCalledTimes(1);
  });

  it('should fetch rows on page change', () => {
    wrapper.instance().handlePageChange(1);
    expect(props.fetchRows).toHaveBeenCalledWith({
      currentPage: 2,
      perPage: props.defaultPerPage
    });
  });

  it('should fetch rows on page size change', () => {
    wrapper.instance().handlePerPageChange(2);
    expect(props.fetchRows).toHaveBeenCalledWith({
      currentPage: 1,
      perPage: 2
    });
  });

  it('should render sync fetched rows', async () => {
    const { rows, ...rest } = props;
    wrapper = shallow(<PaginatingCollection {...rest} />);
    await wrapper.instance().updateRows();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render async fetched rows', async () => {
    const { rows, ...rest } = props;
    props.fetchRows.mockResolvedValue({ rows, rowCount: rows.length });
    wrapper = shallow(<PaginatingCollection {...rest} />);
    await wrapper.instance().updateRows();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should update the query string if page is present', () => {
    const { defaultPerPage, ...rest } = props;
    rest.location.search = '?page=2&other=cool';
    rest.location.pathname = 'some-pathname';
    rest.updateQueryString = true;
    wrapper = shallow(<PaginatingCollection {...rest} />);
    wrapper.instance().maybeUpdateQueryString();
    expect(props.history.push).toHaveBeenCalledWith('some-pathname?other=cool&page=2&perPage=25');
  });

  it.skip('should update the query string if updateQueryString prop is true', () => {
    props.location.search = '?other=cool';
    props.location.pathname = 'some-pathname';
    props.updateQueryString = true;
    wrapper = shallow(<PaginatingCollection {...props} />);
    wrapper.instance().maybeUpdateQueryString();
    expect(props.history.push).toHaveBeenCalledWith('some-pathname?other=cool&page=1&perPage=10');
  });
});
