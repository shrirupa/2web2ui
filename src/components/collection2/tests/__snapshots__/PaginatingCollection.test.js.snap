// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: PaginatingCollection should render async fetched rows 1`] = `
<React.Fragment>
  <TableCollectionView
    defaultPerPage={5}
    fetchRows={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "currentPage": 1,
              "perPage": 5,
            },
          ],
          Array [
            Object {
              "currentPage": 1,
              "perPage": 5,
            },
          ],
        ],
        "results": undefined,
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={
      Object {
        "pathname": "/",
        "search": "",
      }
    }
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
  <CollectionPagination
    currentPage={1}
    onPageChange={[Function]}
    onPerPageChange={[Function]}
    pageRange={5}
    perPage={5}
    perPageButtons={
      Array [
        10,
        25,
        50,
        100,
      ]
    }
    rowCount={10}
    saveCsv={false}
  />
</React.Fragment>
`;

exports[`Component: PaginatingCollection should render initial rows if provided 1`] = `
<React.Fragment>
  <TableCollectionView
    defaultPerPage={5}
    fetchRows={[MockFunction]}
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={
      Object {
        "pathname": "/",
        "search": "",
      }
    }
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
  <CollectionPagination
    currentPage={1}
    onPageChange={[Function]}
    onPerPageChange={[Function]}
    pageRange={5}
    perPage={5}
    perPageButtons={
      Array [
        10,
        25,
        50,
        100,
      ]
    }
    rowCount={10}
    saveCsv={false}
  />
</React.Fragment>
`;

exports[`Component: PaginatingCollection should render sync fetched rows 1`] = `
<React.Fragment>
  <TableCollectionView
    defaultPerPage={5}
    fetchRows={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "currentPage": 1,
              "perPage": 5,
            },
          ],
          Array [
            Object {
              "currentPage": 1,
              "perPage": 5,
            },
          ],
        ],
        "results": undefined,
      }
    }
    history={
      Object {
        "push": [MockFunction],
      }
    }
    location={
      Object {
        "pathname": "/",
        "search": "",
      }
    }
    rows={
      Array [
        Object {
          "id": 0,
          "name": "row-name-0-0",
        },
        Object {
          "id": 1,
          "name": "row-name-0-1",
        },
        Object {
          "id": 2,
          "name": "row-name-0-2",
        },
        Object {
          "id": 3,
          "name": "row-name-0-3",
        },
        Object {
          "id": 4,
          "name": "row-name-0-4",
        },
        Object {
          "id": 5,
          "name": "row-name-1-5",
        },
        Object {
          "id": 6,
          "name": "row-name-1-6",
        },
        Object {
          "id": 7,
          "name": "row-name-1-7",
        },
        Object {
          "id": 8,
          "name": "row-name-1-8",
        },
        Object {
          "id": 9,
          "name": "row-name-1-9",
        },
      ]
    }
  />
  <CollectionPagination
    currentPage={1}
    onPageChange={[Function]}
    onPerPageChange={[Function]}
    pageRange={5}
    perPage={5}
    perPageButtons={
      Array [
        10,
        25,
        50,
        100,
      ]
    }
    rowCount={10}
    saveCsv={false}
  />
</React.Fragment>
`;
